### PR TITLE
Optimize ColumnString::replicate and prevent memcpySmallAllowReadWriteOverflow15Impl from being optimized to built-in memcpy

### LINKED ISSUE #61074: I found memcpySmallAllowReadWriteOverflow15Impl will call built-in memcpy  because of clang loop-idiom optimization. It is as expected, or an issue?
dc4d586ae2a041a2e992185fc004747

 https://godbolt.org/z/YxM3rc7Eh

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize ColumnString::replicate and prevent memcpySmallAllowReadWriteOverflow15Impl from being optimized to built-in memcpy. Close https://github.com/ClickHouse/ClickHouse/issues/61074.  
ColumnString::replicate speeds up by 2.46x on x86-64. 
 

Benchmark is in https://github.com/bigo-sg/gluten/tree/b9580154365c09fd12057ba4e69cf3db03d1b7c3

Baseline: 
```
-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_replicate<type_string>             3449203 ns      3449054 ns          202
BM_replicate<type_nullable_string>    4547439 ns      4547316 ns          154
``` 

Opt1: Calculate and allocate chars size in advance and avoid resize/push_back being called frequently
```
-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_replicate<type_string>             2957870 ns      2957805 ns          236
BM_replicate<type_nullable_string>    4088386 ns      4088250 ns          169
```

Opt2: Prevent memcpySmallAllowReadWriteOverflow15Impl from being optimized to built-in memcpy, which is slow in current scenerio
```
-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_replicate<type_string>             1401663 ns      1401607 ns          504
BM_replicate<type_nullable_string>    2500373 ns      2500282 ns          279
``` 

Reference:  https://www.zhihu.com/question/647466901/answer/3422276125

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 89ae39e598a481dbb2c610ee7dca1fc7272517b7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61075/89ae39e598a481dbb2c610ee7dca1fc7272517b7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by taiyang-li** [comment]
Let's wait for performance-test report in CI.

File: src/Common/memcpySmall.h [hunk 1]
```diff
@@ -49,6 +49,9 @@ namespace detail
             dst += 16;
             src += 16;
             n -= 16;
+
+            /// Avoid clang loop-idion optimization, which transforms _mm_storeu_si128 to built-in memcpy
```
Comment (alexey-milovidov) [inline src/Common/memcpySmall.h:53]: Idion -> idiom

**Comment by taiyang-li** [comment]
clickhouse build in https://s3.amazonaws.com/clickhouse-test-reports/61075/d00122384d0645da219e72ef2c1285f7bf521f4f/clickhouse_build_check/report.html  failed, do you know why @nickitat

**Comment by taiyang-li** [comment]
Tests based on sql: 
```
with 'clickhouse' as str
select arrayJoin(range(number % 10)),  materialize(str) from numbers(100000000) format Null;  

Before: 
0 rows in set. Elapsed: 5.450 sec. Processed 100.00 million rows, 800.00 MB (18.35 million rows/s., 146.78 MB/s.)
Peak memory usage: 16.63 MiB.

After: 
0 rows in set. Elapsed: 2.490 sec. Processed 100.00 million rows, 800.00 MB (40.16 million rows/s., 321.28 MB/s.)
Peak memory usage: 10.18 MiB.
```

**Review by nickitat — APPROVED** [review]

**Comment by taiyang-li** [comment]
Notice that `hashjoin with large output` speeds up by 1.19x

**Comment by alexey-milovidov** [comment]
@taiyang-li Please merge with master.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**